### DISCUSSION TITLE
Hardware accelerated video transcoding with multiple worker nodes with dissimilar GPUs/APIs/vendors does not work

### DISCUSSION THREAD

**Discussion by skelefour** [immich-app/immich#12925] [Q&A] [answered]
### The bug

I'm unsure if this is the expected behavior, but while scaling my installation across a second device, it appears as of now, to be **impossible to mix hardware acceleration APIs between worker nodes.**

### The OS that Immich Server is running on

Debian 12

### Version of Immich Server

v1.115.0

### Version of Immich Mobile App

n/a

### Platform with the issue

- [X] Server
- [ ] Web
- [ ] Mobile

### Your docker-compose.yml content

```YAML
standard
```

### Your .env content

```Shell
standard
```

### Reproduction steps

Hardware:
    1. Main server with an Intel GPU supporting QuickSync Video
    2. A separate "immich_server" worker node on a second machine with an Nvidia GPU supporting NVENC

With the webui configured at:
    Administration > Server Settings > Video Transcode Settings
to use the QuickSync Video for transcoding, **immich expects and requires the additional worker node to also use QSV.**

A likely workaround would be to use the VAAPI backend instead, as it is more widely supported across GPU vendors, however, per the docs, use of the hardware specific APIs is generally preferred. The docs also currently state immich does not support hardware _decoding_ with VAAPI, so CPU load would also increase somewhat.

Ideally, individual configuration of hardware transcoding would be possible for each worker, allowing each worker to utilize their preferred acceleration API.

Relevant log from worker node equipped with Nvidia GPU attached below.

### Relevant log output

```shell
immich_server            | [Nest] 7  - 09/25/2024, 6:19:29 AM   ERROR [Microservices:MediaService] An error occurred while configuring transcoding options: Error: No QSV device found
immich_server            | [Nest] 7  - 09/25/2024, 6:19:29 AM   ERROR [Microservices:MediaService] An error occurred while configuring transcoding options: Error: No QSV device found
immich_server            | [Nest] 7  - 09/25/2024, 6:19:29 AM   ERROR [Microservices:MediaService] An error occurred while configuring transcoding options: Error: No QSV device found
immich_server            | [Nest] 7  - 09/25/2024, 6:19:29 AM   ERROR [Microservices:MediaService] An error occurred while configuring transcoding options: Error: No QSV device found
immich_server            | [Nest] 7  - 09/25/2024, 6:19:30 AM   ERROR [Microservices:MediaService] An error occurred while configuring transcoding options: Error: No QSV device found
immich_server            | [Nest] 7  - 09/25/2024, 6:19:30 AM   ERROR [Microservices:MediaService] An error occurred while configuring transcoding options: Error: No QSV device found
immich_server            | [Nest] 7  - 09/25/2024, 6:19:30 AM   ERROR [Microservices:MediaService] An error occurred while configuring transcoding options: Error: No QSV device found
immich_server            | [Nest] 7  - 09/25/2024, 6:19:30 AM   ERROR [Microservices:MediaService] An error occurred while configuring transcoding options: Error: No QSV device found
immich_server            | [Nest] 7  - 09/25/2024, 6:19:30 AM   ERROR [Microservices:MediaService] An error occurred while configuring transcoding options: Error: No QSV device found
immich_server            | [Nest] 7  - 09/25/2024, 6:19:30 AM   ERROR [Microservices:MediaService] An error occurred while configuring transcoding options: Error: No QSV device found
immich_server            | [Nest] 7  - 09/25/2024, 6:19:31 AM   ERROR [Microservices:MediaService] An error occurred while configuring transcoding options: Error: No QSV device found
immich_server            | [Nest] 7  - 09/25/2024, 6:19:31 AM   ERROR [Microservices:MediaService] An error occurred while configuring transcoding options: Error: No QSV device found
immich_server            | [Nest] 7  - 09/25/2024, 6:19:31 AM   ERROR [Microservices:MediaService] An error occurred while configuring transcoding options: Error: No QSV device found
```

**Comment by bo0tzz** [comment, marked as answer]
The UI config is currently not built with heterogenous workers in mind, though that may change in the future. For now we recommend giving each instance its own [config file](https://immich.app/docs/install/config-file) with the settings it needs. While that should work fine, we haven't specifically tested it, so if it causes any weird issues please report those.

**Reply by skelefour** [reply]
Thanks for your reply.

In my brief testing, providing a separate immich-config.json to the worker node **does** seem to resolve this. I was able to confirm QSV usage on the server and NVENC on the worker node, and have experienced no ill effect.

To achieve this, I exported a configuration file from the webUI set to use NVENC, then added:

```
IMMICH_CONFIG_FILE=/immich-config.json
```

to the worker node's .env file, and attached this config to the container by adding a docker volume mount on the worker:

```
volumes:
      - ${UPLOAD_LOCATION}:/usr/src/app/upload
      - /etc/localtime:/etc/localtime:ro
      - ./immich-config.json:/immich-config.json:ro
```

I also tried mixing target video codecs, which appeared to work as well, with QSV encoding to VP9 and NVENC to H264.
